File: src/main/java/org/radarbase/management/web/rest/MetaTokenResource.java
package org.radarbase.management.web.rest;


import io.micrometer.core.annotation.Timed;
import org.radarbase.auth.config.Constants;
import org.radarbase.auth.exception.NotAuthorizedException;
import org.radarbase.auth.token.RadarToken;
import org.radarbase.management.domain.MetaToken;
import org.radarbase.management.domain.Subject;
import org.radarbase.management.service.MetaTokenService;
import org.radarbase.management.service.dto.ClientPairInfoDTO;
import org.radarbase.management.service.dto.TokenDTO;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.net.MalformedURLException;
import java.time.Duration;

import static org.radarbase.auth.authorization.Permission.SUBJECT_UPDATE;
import static org.radarbase.auth.authorization.RadarAuthorization.checkPermissionOnSubject;

@RestController
@RequestMapping("/api")
public class MetaTokenResource {

    private static final Logger log = LoggerFactory.getLogger(OAuthClientsResource.class);

    public static final Duration DEFAULT_META_TOKEN_TIMEOUT = Duration.ofHours(1);
    public static final Duration DEFAULT_PERSISTENT_META_TOKEN_TIMEOUT = Duration.ofDays(31);

    @Autowired
    private MetaTokenService metaTokenService;

    /**
     * GET /api/meta-token/:tokenName.
     *
     * <p>Get refresh-token available under this tokenName.</p>
     *
     * @param tokenName the tokenName given after pairing the subject with client
     * @return the client as a {@link ClientPairInfoDTO}
     */
    @GetMapping("/meta-token/{tokenName:" + Constants.TOKEN_NAME_REGEX + "}")
    @Timed
    public ResponseEntity<TokenDTO> getTokenByTokenName(@PathVariable("tokenName") String tokenName)
            throws MalformedURLException {
        log.info("Requesting token with tokenName {}", tokenName);
        return ResponseEntity.ok().body(metaTokenService.fetchToken(tokenName));
    }

    /**
     * DELETE /api/meta-token/:tokenName.
     *
     * <p>Delete refresh-token available under this tokenName.</p>
     *
     * @param tokenName the tokenName given after pairing the subject with client
     * @return the client as a {@link ClientPairInfoDTO}
     */
    @DeleteMapping("/meta-token/{tokenName:" + Constants.TOKEN_NAME_REGEX + "}")
    @Timed
    public ResponseEntity<Void> deleteTokenByTokenName(
            @PathVariable("tokenName") String tokenName, RadarToken token)
            throws NotAuthorizedException {
        log.info("Requesting token with tokenName {}", tokenName);
        MetaToken metaToken = metaTokenService.getToken(tokenName);
        Subject subject = metaToken.getSubject();
        String project = subject.getActiveProject()
                .orElseThrow(() -> new NotAuthorizedException(
                        "Cannot establish authority of subject without active project affiliation."
                ))
                .getProjectName();
        String user = subject.getUser().getLogin();
        checkPermissionOnSubject(token, SUBJECT_UPDATE, project, user);
        metaTokenService.delete(metaToken);
        return ResponseEntity.noContent().build();
    }
}

File: src/main/java/org/radarbase/management/web/rest/MetaTokenResource.kt
package org.radarbase.management.web.rest

import io.micrometer.core.annotation.Timed
import org.radarbase.auth.authorization.EntityDetails
import org.radarbase.auth.authorization.Permission
import org.radarbase.management.security.Constants
import org.radarbase.management.security.NotAuthorizedException
import org.radarbase.management.service.AuthService
import org.radarbase.management.service.MetaTokenService
import org.radarbase.management.service.dto.TokenDTO
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import java.net.MalformedURLException
import java.time.Duration

@RestController
@RequestMapping("/api")
class MetaTokenResource {
    @Autowired
    private val metaTokenService: MetaTokenService? = null

    @Autowired
    private val authService: AuthService? = null

    /**
     * GET /api/meta-token/:tokenName.
     *
     *
     * Get refresh-token available under this tokenName.
     *
     * @param tokenName the tokenName given after pairing the subject with client
     * @return the client as a [ClientPairInfoDTO]
     */
    @GetMapping("/meta-token/{tokenName:" + Constants.TOKEN_NAME_REGEX + "}")
    @Timed
    @Throws(
        MalformedURLException::class,
    )
    fun getTokenByTokenName(
        @PathVariable("tokenName") tokenName: String?,
    ): ResponseEntity<TokenDTO> {
        log.info("Requesting token with tokenName {}", tokenName)
        return ResponseEntity.ok().body(tokenName?.let { metaTokenService!!.fetchToken(it) })
    }

    /**
     * DELETE /api/meta-token/:tokenName.
     *
     *
     * Delete refresh-token available under this tokenName.
     *
     * @param tokenName the tokenName given after pairing the subject with client
     * @return the client as a [ClientPairInfoDTO]
     */
    @DeleteMapping("/meta-token/{tokenName:" + Constants.TOKEN_NAME_REGEX + "}")
    @Timed
    @Throws(
        NotAuthorizedException::class,
    )
    fun deleteTokenByTokenName(
        @PathVariable("tokenName") tokenName: String?,
    ): ResponseEntity<Void> {
        log.info("Requesting token with tokenName {}", tokenName)
        val metaToken = tokenName?.let { metaTokenService!!.getToken(it) }
        val subject = metaToken?.subject
        val project: String =
            subject!!
                .activeProject
                ?.projectName
                ?: throw NotAuthorizedException(
                    "Cannot establish authority of subject without active project affiliation.",
                )
        val user = subject.user!!.login
        authService!!.checkPermission(
            Permission.SUBJECT_UPDATE,
            { e: EntityDetails -> e.project(project).subject(user) },
        )
        metaTokenService?.delete(metaToken)
        return ResponseEntity.noContent().build()
    }

    companion object {
        private val log = LoggerFactory.getLogger(OAuthClientsResource::class.java)

        @JvmField
        val DEFAULT_META_TOKEN_TIMEOUT = Duration.ofHours(1)

        @JvmField
        val DEFAULT_PERSISTENT_META_TOKEN_TIMEOUT = Duration.ofDays(31)
    }
}
